src/core/border_router/routing_manager.cpp

```diff
@@ -373,22 +471,50 @@ bool RoutingManager::ContainsPrefix(const Ip6::Prefix &aPrefix, const Ip6::Prefi
     return ret;
 }
 
-const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void) const
+const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void)
 {
-    const Ip6::Prefix *newOnLinkPrefix = nullptr;
+    const Ip6::Prefix *newOnLinkPrefix      = nullptr;
+    const Ip6::Prefix *smallestOnLinkPrefix = nullptr;
+    uint8_t            onLinkPrefixNum      = 0;
 
-    // We don't evaluate on-link prefix if we are doing
-    // Router Discovery or we have already discovered some
-    // on-link prefixes.
+    // We don't evaluate on-link prefix if we are doing Router Solicitation.
     VerifyOrExit(!mRouterSolicitTimer.IsRunning());
-    if (IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix))
+
+    for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
```

Comment:
Just a suggestion: how about adding `ExternalPrefix &prefix = mDiscoveredPrefixes[i];` in `for` block (I like that we have this in `InvalidateDiscoveredPrefixes` and `AddDiscoveredPrefix` and makes it nicer and easier to read)

Reply:
Sure! I share the same preference but maybe forgot to update the loop here.